### ISSUE TITLE
🐞 Optional env vars cannot be used as env vars

### ISSUE THREAD

**Issue by sagikazarmark** [dagger/dagger#2680]
### What is the issue?

Environment variables marked as optional cannot be used as environment variables in actions. For example, the following plan would fail:

```cue
package main

import (
	"dagger.io/dagger"

	"universe.dagger.io/alpine"
	"universe.dagger.io/bash"
)

dagger.#Plan & {
	client: env: {
		TEST_OPTIONAL?: dagger.#Secret
	}
	actions: {
		test: {
			run: bash.#Run & {
				_image: alpine.#Build & {
					packages: {
						bash: _
					}
				}
				input: _image.output
				script: contents: "test \"$TEST_OPTIONAL\" = foobar"
				if client.env.TEST_OPTIONAL != _|_ {
					env: TEST_OPTIONAL: client.env.TEST_OPTIONAL
				}
			}
		}
	}
}
```

It's not that simple though: if I remove the conditional check, the plan succeeds.

To make matters even more complicated, I tried to submit a failing test: #2679 ....but it doesn't fail.

Based on that, I tried to update the above plan with a snippet from the aformentioned test:

```cue
package main

import (
	"dagger.io/dagger"
	"dagger.io/dagger/core"

	"universe.dagger.io/alpine"
	"universe.dagger.io/bash"
)

dagger.#Plan & {
	client: env: {
		TEST_OPTIONAL?: dagger.#Secret
	}
	actions: {
		image: core.#Pull & {
			source: "alpine:3.15.0@sha256:e7d88de73db3d3fd9b2d63aa7f447a10fd0220b7cbf39803c803f2af9ba256b3"
		}
		test: {
			exec: core.#Exec & {
				input: image.output
				if client.env.TEST_OPTIONAL != _|_ {
					env: TEST_OPTIONAL: client.env.TEST_OPTIONAL
				}
				args: ["sh", "-c", "test $TEST_OPTIONAL = foobar"]
			}
			run: bash.#Run & {
				_image: alpine.#Build & {
					packages: {
						bash: _
					}
				}

				input: _image.output
				script: contents: "test \"$TEST_OPTIONAL\" = foobar"
				if client.env.TEST_OPTIONAL != _|_ {
					env: TEST_OPTIONAL: client.env.TEST_OPTIONAL
				}
			}
		}
	}
}
```

As you might guess, it no longer fails.

### Log output

❯ TEST_OPTIONAL=foobar dagger do -p secrettest.cue test --log-format plain --no-cache
10:55AM INF actions.test.run.script._write | computing
10:55AM INF client.env | computing
10:55AM INF actions.test.run._image._dag."0"._op | computing
10:55AM INF client.env | completed    duration=0s
10:55AM INF actions.test.run.script._write | completed    duration=0s
10:55AM INF actions.test.run._image._dag."0"._op | completed    duration=0s
10:55AM INF actions.test.run._image._dag."1"._exec | computing
10:55AM INF actions.test.run._image._dag."1"._exec | #4 0.081 fetch https://dl-cdn.alpinelinux.org/alpine/v3.15/main/x86_64/APKINDEX.tar.gz
10:55AM INF actions.test.run._image._dag."1"._exec | #4 0.419 fetch https://dl-cdn.alpinelinux.org/alpine/v3.15/community/x86_64/APKINDEX.tar.gz
10:55AM INF actions.test.run._image._dag."1"._exec | #4 0.672 (1/4) Installing ncurses-terminfo-base (6.3_p20211120-r0)
10:55AM INF actions.test.run._image._dag."1"._exec | #4 0.710 (2/4) Installing ncurses-libs (6.3_p20211120-r0)
10:55AM INF actions.test.run._image._dag."1"._exec | #4 0.753 (3/4) Installing readline (8.1.1-r0)
10:55AM INF actions.test.run._image._dag."1"._exec | #4 0.792 (4/4) Installing bash (5.1.16-r0)
10:55AM INF actions.test.run._image._dag."1"._exec | #4 0.909 Executing bash-5.1.16-r0.post-install
10:55AM INF actions.test.run._image._dag."1"._exec | completed    duration=1s
10:55AM INF actions.test.run._exec | computing
10:55AM INF actions.test.run._image._dag."1"._exec | #4 0.912 Executing busybox-1.34.1-r3.trigger
10:55AM INF actions.test.run._image._dag."1"._exec | #4 0.917 OK: 8 MiB in 18 packages
10:55AM INF actions.test.run._exec | completed    duration=100ms
❯ TEST_OPTIONAL=foobar dagger do -p secrettest.cue test --log-format plain --no-cache
10:55AM INF actions.test.run.script._write | computing
10:55AM INF actions.test.run._image._dag."0"._op | computing
10:55AM INF actions.test.run.script._write | completed    duration=0s
10:55AM INF actions.test.run._image._dag."0"._op | completed    duration=0s
10:55AM INF actions.test.run._image._dag."1"._exec | computing
10:55AM INF actions.test.run._image._dag."1"._exec | #4 0.082 fetch https://dl-cdn.alpinelinux.org/alpine/v3.15/main/x86_64/APKINDEX.tar.gz
10:55AM INF actions.test.run._image._dag."1"._exec | #4 0.431 fetch https://dl-cdn.alpinelinux.org/alpine/v3.15/community/x86_64/APKINDEX.tar.gz
10:55AM INF actions.test.run._image._dag."1"._exec | #4 0.674 (1/4) Installing ncurses-terminfo-base (6.3_p20211120-r0)
10:55AM INF actions.test.run._image._dag."1"._exec | #4 0.720 (2/4) Installing ncurses-libs (6.3_p20211120-r0)
10:55AM INF actions.test.run._image._dag."1"._exec | #4 0.769 (3/4) Installing readline (8.1.1-r0)
10:55AM INF actions.test.run._image._dag."1"._exec | #4 0.814 (4/4) Installing bash (5.1.16-r0)
10:55AM INF actions.test.run._image._dag."1"._exec | #4 0.933 Executing bash-5.1.16-r0.post-install
10:55AM INF actions.test.run._image._dag."1"._exec | completed    duration=1s
10:55AM INF actions.test.run._exec | computing
10:55AM INF actions.test.run._image._dag."1"._exec | #4 0.937 Executing busybox-1.34.1-r3.trigger
10:55AM INF actions.test.run._image._dag."1"._exec | #4 0.941 OK: 8 MiB in 18 packages
10:55AM ERR actions.test.run._exec | failed: process "bash --norc -e -o pipefail /bash/scripts/run.sh" did not complete successfully: exit code: 1
    duration=100ms
10:55AM FTL system | failed to execute plan: task failed: actions.test.run._exec: process "bash --norc -e -o pipefail /bash/scripts/run.sh" did not complete successfully: exit code: 1

### Steps to reproduce

See the above plans.

### Dagger version

0.2.19

### OS version

macOS

**Comment by helderco** [comment]
This is one of those head scratchers from CUE's evaluator. Try this:

```cue
package main

import (
	"dagger.io/dagger"
	"universe.dagger.io/alpine"
	"universe.dagger.io/bash"
)

dagger.#Plan & {
	client: env: {
		TEST_OPTIONAL?: dagger.#Secret
	}
	actions: {
		test: {
			run: bash.#Run & {
				_image: alpine.#Build & {
					packages: bash: _
				}
				input: _image.output
				script: contents: "test $TEST_OPTIONAL = foobar"
				env: {
					if client.env.TEST_OPTIONAL != _|_ {
						TEST_OPTIONAL: client.env.TEST_OPTIONAL
					}
				}
			}
		}
	}
}
```

Notice how the condition has moved inside `env`. We advise this in https://docs.dagger.io/1226/coding-style#scope-conditions-and-loops-as-locally-as-possible. I learned it from Marcel in an issue, along with an explanation. We should dig that out.

**Comment by helderco** [comment]
I think I saw it here: https://github.com/cue-lang/cue/issues/1130#issuecomment-954780228

> In general, it is a good idea to move comprehensions to the most deeply nested struct to which they apply.

**Comment by sagikazarmark** [comment]
Thanks @helderco ! The above is just an example that demonstrates the problem, but i can reproduce it with custom definitions as well. For example:

```cue
upload: codecov.#Upload & {
    source: _source.output
    file:   "coverage.out"

    if client.env.CODECOV_TOKEN != _|_ {
        token: client.env.CODECOV_TOKEN
    }
}
```

Then in the codecov package:

```cue
#Upload: {
	// Source code
	source: dagger.#FS

	// Coverage files
	file: string

	// Codecov token (required for local runs and private repos)
	token?: dagger.#Secret

	_image: #Image

	_sourcePath: "/src"

	docker.#Run & {
		input: *_image.output | docker.#Image
		command: {
			name: "codecov"
			flags: {
				"--file": file
			}
		}
		env: {
			if token != _|_ {
				CODECOV_TOKEN: token
			}
		}
		workdir: _sourcePath
		mounts: {
			"source": {
				dest:     _sourcePath
				contents: source
			}
		}
	}
}
```

This doesn't work either and I can't move the condition down.

Funny thing is, when the env var is set and I remove the condition it just works. So I guess you are right that it has to do with when it gets evaluated.

**Comment by sagikazarmark** [comment]
The following seems to be working:

```cue
_env: client.env

if _env.CODECOV_TOKEN != _|_ {
    token: _env.CODECOV_TOKEN
}
```

**Comment by sagikazarmark** [comment]
I thought the above would solve the problem, but as soon as I add any additional env vars to the client it fails again:

```cue
	client: env: {
		GITHUB_ACTIONS:    string | *""
		GITHUB_HEAD_REF:   string | *""
		GITHUB_REF:        string | *""
		GITHUB_REPOSITORY: string | *""
		GITHUB_RUN_ID:     string | *""
		GITHUB_SHA:        string | *""
		GITHUB_WORKFLOW:   string | *""
		CODECOV_TOKEN?:    dagger.#Secret
	}

```

```
7:30PM INF actions.check.test.go."1.18"._coverage._image._dag."7".script._write | computing
7:30PM INF actions.check.test.go."1.18"._test._image._dag."0"._op | computing
7:30PM INF client.env | computing
7:30PM INF client.filesystem.".".read | computing
7:30PM INF actions.check.test.go."1.18"._coverage._image._dag."0"._op | computing
7:30PM INF client.env | completed    duration=0s
7:30PM INF actions.check.test.go."1.18"._coverage._image._dag."7".script._write | completed    duration=0s
7:30PM INF client.filesystem.".".read | completed    duration=0s
7:30PM INF actions.check.test.go."1.18"._coverage._image._dag."0"._op | completed    duration=0s
7:30PM INF actions.check.test.go."1.18"._coverage._env | computing
7:30PM ERR actions.check.test.go."1.18"._coverage._env | failed: GITHUB_ACTIONS: unexpected concrete value, please use a type or set a default
    duration=0s
7:30PM INF actions.check.test.go."1.18"._coverage._image._dag."1"._exec | computing
7:30PM ERR actions.check.test.go."1.18"._test._image._dag."0"._op | cancelled    duration=700ms
7:30PM ERR actions.check.test.go."1.18"._coverage._image._dag."1"._exec | cancelled    duration=0s
7:30PM FTL system | failed to execute plan: task failed: actions.check.test.go."1.18"._coverage._env: GITHUB_ACTIONS: unexpected concrete value, please use a type or set a default
```

**Comment by sagikazarmark** [comment]
I also tried just passing in the optional env var to the token parameter:

```cue
token: client.env.CODECOV_TOKEN
```

Then in the Upload definition:

```
		env: {
			if token != _|_ {
				CODECOV_TOKEN: token
			}
		}
```

That results in the following error: 

```
7:34PM FTL system | failed to execute plan: task failed: actions.check.test.go."1.18"._coverage._exec: actions.check.test.go."1.18"._coverage._exec.env: cycle error
```

**Comment by verdverm** [comment]
We just talked about this issue on the CUE Slack (https://cuelang.slack.com/archives/CLT3ULF6C/p1656418916232699)

tl;dr, non-regular fields (like optional) have different semantics than regular fields. In particular they are not seen during looping and seemingly if (guards)

That being said, this may change with recent comprehension work on tip, but not released yet.

We could potentially test if the changes fix this with https://github.com/hofstadter-io/harmony-cue. I'm pretty sure I ran some tests where I built and tested dagger with an arbitrary commit from CUE.

**Comment by verdverm** [comment]
Your `dagger.#Secret` will never equal bottom, since it is defined as a struct with fields already, thus it always has a value and the `?` is more like a no-op (though that is an oversimplification beyond this limited context)

**Comment by sagikazarmark** [comment]
@verdverm that's not what I experience.

When I make `token` a non-optional field in the above example and the env var is not set, dagger complains about the token not being set.

That makes me believe that the equality check works for the CODECOV_TOKEN env var. (It still needs the line above though, otherwise it indeed doesn't work)

The documentation also contains examples for optional secret env vars: https://docs.dagger.io/1203/client/#environment-variables

**Comment by helderco** [comment]
The optional sign isn't handled entirely in CUE. It's used by the engine to decide if the plan should fail or not if the var isn't set, unless it's a string and there's a default value.

**Comment by helderco** [comment]
@sagikazarmark If you have a simple value instead of a struct:

```cue
upload: codecov.#Upload & {
    if client.env.CODECOV_TOKEN != _|_ {
        token: client.env.CODECOV_TOKEN
    }
}
```
...you can still scope it more locally, it's the same reason as my first comment:

```cue
upload: codecov.#Upload & {
    token: {
        if client.env.CODECOV_TOKEN != _|_ {
            client.env.CODECOV_TOKEN
        }
    }
}
```

Notice the embedding (only value, no field):

```cue
a: { b }

// equal
a: b
```

**Comment by helderco** [comment]
@sagikazarmark You also said that as soon as you added other vars it stopped working: https://github.com/dagger/dagger/issues/2680#issuecomment-1159781477

But looking at the errors you see this:

```console
actions.check.test.go."1.18"._coverage._env | failed: GITHUB_ACTIONS: unexpected concrete value, please use a type or set a default
```

Without seeing your plan, this makes me think that you have this:

```cue
actions: check: test: go: "1.18": _coverage: _env: client.env
```

You can't send `client.env` into actions because that'll copy the core action and run it again but this time with concrete values if you have some of the vars actually set in the host. In this case, it's complaining about `GITHUB_ACTIONS` being concrete.

If I'm off the mark here, let me know. I'd need to see more of the plan to help you understand but it should be something like that.

There's actually a way to pass the entire `client.env` into an action but you have to exclude the `$dagger` field from there which is what makes dagger run it in the engine:

```
actions: _clientEnv: {
	for k, v in client.env {
		if k != "$dagger" {
			(k): v
		}
	}
}
```

Then you can pass this `_clientEnv` along.

**Comment by sagikazarmark** [comment]
> Problem then is that if var isn't set, you're setting the `token` field to something (not optional... it will give you an _invalid secret_ error). But you can do this instead:
> 
> ```cue
> #Upload: {
> 	token: *null | dagger.#Secret
> 
> 	docker.#Run & {
> 		env: {
> 			if token != null {
> 				CODECOV_TOKEN: token
> 			}
> 		}
> 		...
> 	}
> 	...
> }
> ```

Thanks @helderco! Indeed, this seems to be working, but I wonder what the added value of the optional feature is if you need "workarounds" like this to use it. I mean, falling back to an empty string and comparing with that sounds like a much cleaner solution at this point. This is why I said this feature is more confusing than useful at the moment.

> If I'm off the mark here, let me know. I'd need to see more of the plan to help you understand but it should be something like that.

You are absolutely right and makes sense. This was only one of the solutions I tried to figure out why optional env doesn't work for me.

**Comment by helderco** [comment]
> I wonder what the added value of the optional feature is if you need "workarounds" like this to use it.

The problem is you need a secret, which isn't the same as a string. So doesn't make sense to use a string as a default to a secret. 

I tried using `null` first, but I found it more complicated. You can find an explanation in (see _What else was considered?_):
- https://github.com/dagger/dagger/pull/2370

I could allow a default secret, which would have to come from somewhere else, but wouldn't help you in your case.

I don't see that you need a workaround. It's just the difference of using a _string_ primitive vs a _struct_ (secret), and trying to follow Marcel's (CUE creator) recommendation of nesting comprehensions as deep as possible due to how CUE's evaluator works.

If it were some other default instead of the optional (`?`) field, you still need a comprehension to toggle behavior.

**Comment by helderco** [comment]
Also, from that PR you can see that even for strings, an empty value as a default isn't enough in some cases because an env var with an empty value can be a valid config. So you need to tell it apart from not being set. 

Marking as optional then helps in both cases.

**Comment by helderco** [comment]
From your description:

> To make matters even more complicated, I tried to submit a failing test: https://github.com/dagger/dagger/pull/2679 ....but it doesn't fail.
>
> Based on that, I tried to update the above plan with a snippet from the aformentioned test:
>
> [...]
>
> As you might guess, it no longer fails.

The difference in both is that you're seeing this problem (the need to nest deeper) in `bash.#Run` (**`docker.#Run`** in particular), but not in `core.#Exec`. Packages in _universe_ are higher level abstractions on top of core actions and usually contain a bit more logic/dynamism in order to be easier to use. It may also change the order of things for the evaluator in the process though.

I've stripped `docker.#Run` to the bone marrow and still can't seem to pass around it. Just passing `env` along seems to cause this issue:

```cue
docker.#Run & {
    env: {}

    _exec: core.#Exec & {
        "env": env
    }
}
```

**Comment by helderco** [comment]
Ok, this issue no longer happens in CUE 0.4.3. Yet another reason we need the upgrade:

- #2182
